server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -600,6 +604,15 @@ public void onFailure(Exception e) {
         if (newRouting.equals(currentRouting) == false) {
             indexEventListener.shardRoutingChanged(this, currentRouting, newRouting);
         }
+
+        if (indexSettings.isSoftDeleteEnabled() && useRetentionLeasesInPeerRecovery == false) {
+            if (getPeerRecoveryRetentionLeases().size() >= routingTable.size()) {
+                logger.debug("turn off the translog retention for the replication group {} " +
+                    "as it starts using retention leases exclusively in peer recoveries", shardId);
+                useRetentionLeasesInPeerRecovery = true;
+                onSettingsChanged();
```

Comment:
do we need to call onSettingsChanged also when peer recovery retention leases are updated?

Reply:
We should, but here we do not have the routing table unless we also update it on the replica mode.